### PR TITLE
26 07 2022

### PR DESCRIPTION
**Pull request by SebastianBalmus**

### CODE REVIEW

File: cinema/admin.py [hunk 1]
```diff
@@ -1,3 +1,73 @@
 from django.contrib import admin
+from .models import *
```
Comment (maxinasa) [inline cinema/admin.py]: While this works, it's not the best approach to import everything.
We might not get everything we need or even more than we need with it.

File: cinema/forms.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super(ContactForm, self).__init__(*args, **kwargs)
+        self.fields['city'].required = False
+        self.fields['cinema'].required = False
+
+    def clean(self):
+        super(ContactForm, self).clean()
+
+        if not is_only_char(self.cleaned_data['name']):
```
Comment (maxinasa) [inline cinema/forms.py]: Maybe isalpha would have sufficed?

File: cinema/migrations/0007_schedule_cinema.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('cinema', '0006_schedule'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='schedule',
+            name='cinema',
+            field=models.ForeignKey(default=1, on_delete=django.db.models.deletion.CASCADE, to='cinema.cinema'),
```
Comment (maxinasa) [inline cinema/migrations/0007_schedule_cinema.py:17]: The default=1 on the FK looks suspicious to me. Do you remember why is this needed?

File: cinema/models.py [hunk 4]
```diff
@@ -19,6 +24,10 @@ def __str__(self):
     def imdb_link(self):
         return 'https://www.imdb.com/title/' + self.imdb_id.__str__()
```
Comment (maxinasa) [inline cinema/models.py]: We should avoid calling these dunder methods unless needed. Maybe str(self.imdb_id) would have also worked?

File: cinema/models.py [hunk 5]
```diff
@@ -19,6 +24,10 @@ def __str__(self):
...
 
+    @cached_property
+    def embed_trailer(self):
+        return self.trailer_link.__str__().replace('watch?v=', 'embed/')
```
Comment (maxinasa) [inline cinema/models.py]: The same about __str__. If this is a CharField do we really need to make the explicit conversion to access the replace method?

File: cinema/models.py [hunk 6]
```diff
@@ -41,10 +50,73 @@ def __str__(self):
...
+        show_all=False,
+        auto_choose=True,
+        sort=True,
+    )
+    playing_time = models.DateTimeField()
+
+    @cached_property
+    def end_time(self):
+        end_time = self.playing_time
+        parsed_time = re.findall(r'\d+[ ]?', self.movie.length)
```
Comment (maxinasa) [inline cinema/models.py:99]: Maybe it would help to make the movie length field an integer, to store the minutes in it?

File: cinema/models.py [hunk 7]
```diff
@@ -41,10 +50,73 @@ def __str__(self):
...
+            end_hour += 1
+            end_minute -= 60
+
+        if end_hour > 23:
+            end_time = self.playing_time + timedelta(days=1)
+            end_hour -= 24
+
+        end_time = end_time.replace(
+            hour=end_hour,
+            minute=end_minute,
```
Comment (maxinasa) [inline cinema/models.py]: Probably datetime.timedelta would help here.

File: cinema/models.py [hunk 8]
```diff
@@ -41,10 +50,73 @@ def __str__(self):
...
+
+    def clean(self):
+
+        for obj in Schedule.objects.all():
+            if (
+                    self.cinema.name == obj.cinema.name
+                    and self.hall.name == obj.hall.name
+                    and self.playing_time <= obj.end_time
+                    and self.end_time >= obj.playing_time
+            ):
```
Comment (maxinasa) [inline cinema/models.py]: Not sure what this method is used for but you can do a filter instead of looping over all the existing objects and checking each obj value.
```suggestion
            ):Schedule.objects.filter(cinema__name=self.cinema.name, hall__name=self.hall.name, end_time=self.playing_time, playing_time=self.end_time)
            
```
Also, a unique_together (or even better, UniqueConstraint) might help to avoid having the same combination of cinema, hall, movie and playing_time in the DB.

File: cinema/models.py [hunk 9]
```diff
@@ -41,10 +50,73 @@ def __str__(self):
...
+                raise ValidationError('There is a movie playing in that interval')
+
+        if self.playing_time <= timezone.now():
+            raise ValidationError('You can\'t schedule a movie in the past!')
```
Comment (maxinasa) [inline cinema/models.py:111]: Maybe the save method is more suited for this validation?
AFAIK the clean method is not called by default.

File: cinema/views.py [hunk 10]
```diff
@@ -1,7 +1,43 @@
...
+            return render(request, 'cinema/home.html')
+    else:
+        contact_form = ContactForm()
+
+    context = {'form': contact_form}
+    return render(request, 'cinema/contact.html', context)
+
+
+# Currently playing movies
+def currently_playing(request):
```
Comment (maxinasa) [inline cinema/views.py:33]: Indeed, probably we don't need rate limiting for this view.

File: cinema/views.py [hunk 11]
```diff
@@ -1,7 +1,43 @@
...
+    movie_list = Movie.objects.filter(
+        schedule__isnull=False,
+        schedule__playing_time__lte=timezone.now()+timezone.timedelta(days=6),
+        schedule__playing_time__gt=timezone.now()
```
Comment (maxinasa) [inline cinema/views.py:37]: This is good, but just as a side note, these fields also support a "range" lookup.

File: silverscreen/settings.py [hunk 12]
```diff
@@ -113,7 +117,7 @@
 
 LANGUAGE_CODE = 'en-us'
 
-TIME_ZONE = 'UTC'
+TIME_ZONE = 'EET'
```
Comment (maxinasa) [inline silverscreen/settings.py:120]: Any reason why this was chosen?
You can use Europe/Bucharest to view the playing times in our timezone in the admin pages, for example.

**Comment by maxinasa** [comment]
Way too hard to review these many files. :)

File: cinema/models.py (hunk 6 above)
Comment (SebastianBalmus) [reply cinema/models.py:99]: The actual movie length field format looks better to me. Also, this is the standard format for IMDb, so if I want to do some scraping at some point, it will be easier for me.

**Review by maxinasa — APPROVED** [review]
